File: Data Structures & Algorithms/combination-target-sum/submission-8.cpp
class Solution {
public:
vector<vector<int>> ans;
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> curr;
        sort(nums.begin(), nums.end());
        dfs(curr, 0, target, nums);
        return ans;
    }
    void dfs(vector<int>& curr, int i, int target, vector<int>& nums){
        if(target == 0)
            ans.push_back(curr);
        for(int j = i ; j<nums.size(); ++j){
            if(nums[j] > target){
                return;
            }

            curr.push_back(nums[j]);
            dfs(curr, j, target-nums[j], nums);
            curr.pop_back();
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-8.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        res, curr = [], []
        nums.sort()
        def dfs(curr, i, target):
            if target == 0:
                res.append(curr.copy())
                return
            if i == len(nums) or target < nums[i]:
                return
            curr.append(nums[i])
            dfs(curr, i+1, target-nums[i])
            while i <len(nums)-1 and nums[i] == nums[i+1]:
                i+=1
            curr.pop()
            dfs(curr, i+1, target)
        dfs(curr, 0, target)
        return res